src/sentry/api/endpoints/project_rules.py

```diff
@@ -44,66 +48,219 @@ def pre_save_rule(instance, sender, *args, **kwargs):
+        if existing_actions and current_actions:
+            match_results.key_matches = self._compare_lists_of_dicts(
+                keys_to_ignore=["uuid"], list1=existing_actions, list2=current_actions
+            )
+
+        return match_results
+
+    @classmethod
+    def _compare_lists_of_dicts(
+        cls,
+        keys_to_ignore: list[str],
+        list1: list[dict[Any, Any]] | None = None,
+        list2: list[dict[Any, Any]] | None = None,
+    ) -> bool:
+        if list1 is None or list2 is None:
+            return False
+
+        if len(list1) != len(list2):
+            return False
+
+        for i, left in enumerate(list1):
+            right = list2[i]
+            raw_left = {k: v for k, v in left.items() if k not in keys_to_ignore}
+            raw_right = {k: v for k, v in right.items() if k not in keys_to_ignore}
+
+            # TODO (Yash): This code commented below is the corrected logic which accounts for bad key values.
+            # clean_left = cls._get_clean_actions_dict(raw_left)
+            # clean_right = cls._get_clean_actions_dict(raw_right)
+            # if clean_left != clean_right:
+            #     return False
+            """
+            This is a bug in the current logic.
+            When comparing DB values to serialized values, the values that are `None` are not properly converted to
+            empty strings.
+            This means we end up incorrectly evaluating the actions aren't the same, when they actually are.
+            """
+            if raw_left != raw_right:
+                return False
+
+        return True
+
+    @classmethod
+    def _get_clean_actions_dict(cls, actions_dict: dict[any, any]) -> dict[any, any]:
+        """
+        Returns a dictionary where None is substituted with empty string to help compare DB values vs serialized values
+        """
+        cleaned_dict = {}
+        for k, v in actions_dict.items():
+            cleaned_dict[k] = "" if v is None else v
+
+        return cleaned_dict
+
+    def find_duplicate(self) -> Rule | None:
+        """
+        Determines whether specified rule already exists, and if it does, returns it.
+        """
+        existing_rules = Rule.objects.exclude(id=self._rule_id).filter(
+            project__id=self._project_id, status=ObjectStatus.ACTIVE
+        )
+        for existing_rule in existing_rules:
+            keys_checked = 0
+            keys_matched = 0
+            for key_to_check in self._keys_to_check:
+                func = self._get_func_to_call(key_to_check=key_to_check)
+                results: MatcherResult = func(
+                    existing_rule=existing_rule, key_to_check=key_to_check
+                )
+                if results.has_key:
+                    keys_checked += 1
+                    if results.key_matches:
+                        keys_matched += 1
+
+            if keys_checked > 0 and keys_checked == keys_matched:
+                return existing_rule
```

Comment:
Adding some logging here would help prevent from having a bugs like what you mentioned for long.

Reply:
Let's add the logging in the follow up